### ISSUE TITLE
Twisted's Failure fake '_Frame' object lacks f_locals

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#4045]
|<img alt="sidnei's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| sidnei reported|
|-|-|
|Trac ID|trac#4045|
|Type|enhancement|
|Created|2009-09-29 16:31:52Z|

Upon updating Twisted and zope.testing on a known Zope 3-based project, a new failure showed up on a test that was exercising the behavior of trial-based unittests as run by zope.testing's testrunner.

Turns out that a failure wasn't properly being reported due to zope.testing's custom format_traceback function trying to access the 'f_locals' attribute of a frame object (twisted.python.failure._Frame) from the handed over traceback object (an instance of twisted.python.failure._Traceback). 

Of course the code from zope.testing is brittle enough to fall over on its back due to this missing attribute, but there's no apparent reason for Twisted's fake Frame object not provide the said 'f_locals' attribute other than perhaps memory issues (http://bugs.python.org/issue6116).

Attachments:

* [f_locals.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/b49/b49a202cd695007ac4a3749862f273f2e5dde766/ede52ced802a94ec0ffc5a750c05d0e6cc3366c3.patch) (2064 bytes) - added by alanfranzoni on 2010-06-07 21:40:24Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__4045 4045
type__enhancement enhancement
reporter__sidnei sidnei
priority__normal normal
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1254241912000000 1254241912000000
changetime__1277897207000000 1277897207000000
version__None None
owner__ 

```
</details>

**Comment by twisted-trac** [comment]
|<img alt="alanfranzoni's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_1"></a>|alanfranzoni removed owner|
|-|-|

I realize that probably _Frame objects, and other fake objects created by twisted failure infrastructure for exception delivery, will NEVER support the full python sync exception API; most objects got ton of attributes.

By the way zc.buildout and its test runner is gaining momentum in the community, with a point; it's getting increasingly easy to have quickly reproducibile builds and separate work environments in python, something that Java got from years. Hence I think that this small patch (which I would call a feature) deserves its place in twisted trunk.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_2"></a>|@exarkun commented|
|-|-|

... but zope.testing works without `f_locals` now right?

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_3"></a>|@exarkun commented|
|-|-|

(In [29478]) Branching to 'fakeframe-f_locals-4045'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun commented|
|-|-|

(In [29479]) Apply f_locals.patch

refs [#4045](https://github.com/twisted/twisted/issues/4045)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun set owner to @exarkun<br>@exarkun set status to `assigned`|
|-|-|

Alright, well, I have no idea if this will actually help zope.testing or not, but it still sorta makes sense, particularly since `f_globals` is already there.

This basically looks fine.  There are a couple coding standard issues that I'll fix up, then I'll merge the branch.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun set status to `closed`|
|-|-|

(In [29494]) Merge fakeframe-f_locals-4045

Author: alanfranzoni
Reviewer: exarkun
Fixes: [#4045](https://github.com/twisted/twisted/issues/4045)

Add an `f_locals` attribute to the fake frame instances used by
`twisted.python.failure` when real frames aren't available.

This doesn't convey any information about locals, but it might
help tools expecting real frames not break.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
